File: lib/src/models/call_notification_payload.dart
import 'dart:convert';

class CallNotificationPayload {
  const CallNotificationPayload({
    required this.callId,
    required this.roomId,
    required this.callerId,
    required this.callerName,
    required this.receiverId,
    required this.type,
    this.avatarUrl,
    this.mediaType,
    this.metadata = const {},
  });

  final String callId;
  final String roomId;
  final String callerId;
  final String callerName;
  final String receiverId;
  final String type;
  final String? avatarUrl;
  final String? mediaType;
  final Map<String, dynamic> metadata;

  factory CallNotificationPayload.fromJson(Map<String, dynamic> json) {
    return CallNotificationPayload(
      callId: json['callId'] as String? ?? json['roomId'] as String,
      roomId: json['roomId'] as String,
      callerId: json['callerId'] as String,
      callerName: json['callerName'] as String,
      receiverId: json['receiverId'] as String,
      type: json['type'] as String,
      avatarUrl: json['avatarUrl'] as String?,
      mediaType: json['mediaType'] as String?,
      metadata: Map<String, dynamic>.from(
        json['metadata'] is Map
            ? json['metadata'] as Map
            : json['metadata'] is String
                ? _tryDecode(json['metadata'] as String)
                : const {},
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'callId': callId,
      'roomId': roomId,
      'callerId': callerId,
      'callerName': callerName,
      'receiverId': receiverId,
      'type': type,
      'avatarUrl': avatarUrl,
      'mediaType': mediaType,
      'metadata': metadata,
    }..removeWhere((_, value) => value == null);
  }

  static Map<String, dynamic> _tryDecode(String value) {
    try {
      return Map<String, dynamic>.from(
        value.isEmpty ? const {} : (jsonDecode(value) as Map<String, dynamic>),
      );
    } catch (_) {
      return const {};
    }
  }
}

File: android/src/main/kotlin/com/example/call_notification_sdk/model/CallNotificationPayload.kt
package com.example.call_notification_sdk.model

import android.os.Bundle
import org.json.JSONObject

data class CallNotificationPayload(
    val callId: String,
    val roomId: String,
    val callerId: String,
    val callerName: String,
    val receiverId: String,
    val type: String,
    val avatarUrl: String?,
    val mediaType: String?,
    val metadata: Map<String, Any?>,
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "callId" to callId,
        "roomId" to roomId,
        "callerId" to callerId,
        "callerName" to callerName,
        "receiverId" to receiverId,
        "type" to type,
        "avatarUrl" to avatarUrl,
        "mediaType" to mediaType,
        "metadata" to metadata,
    )

    fun toBundle(): Bundle = Bundle().apply {
        putString("callId", callId)
        putString("roomId", roomId)
        putString("callerId", callerId)
        putString("callerName", callerName)
        putString("receiverId", receiverId)
        putString("type", type)
        putString("avatarUrl", avatarUrl)
        putString("mediaType", mediaType)
        putString("metadata", JSONObject(metadata).toString())
    }

    companion object {
        private fun Map<String, String>.requireString(key: String): String {
            return this[key]?.takeIf { it.isNotEmpty() }
                ?: throw IllegalArgumentException("Missing required key: $key")
        }

        fun fromData(data: Map<String, String>): CallNotificationPayload {
            val metadata = data["metadata"]?.let { raw ->
                runCatching {
                    val json = JSONObject(raw)
                    json.keys().asSequence().associateWith { key -> json.opt(key) }
                }.getOrDefault(emptyMap())
            } ?: emptyMap()

            return CallNotificationPayload(
                callId = data["callId"] ?: data.requireString("roomId"),
                roomId = data.requireString("roomId"),
                callerId = data.requireString("callerId"),
                callerName = data.requireString("callerName"),
                receiverId = data.requireString("receiverId"),
                type = data.requireString("type"),
                avatarUrl = data["avatarUrl"],
                mediaType = data["mediaType"],
                metadata = metadata,
            )
        }

        fun fromBundle(bundle: Bundle): CallNotificationPayload {
            val metadata = bundle.getString("metadata")?.let { raw ->
                runCatching {
                    val json = JSONObject(raw)
                    json.keys().asSequence().associateWith { key -> json.opt(key) }
                }.getOrDefault(emptyMap())
            } ?: emptyMap()

            return CallNotificationPayload(
                callId = bundle.getString("callId") ?: bundle.getString("roomId")!!,
                roomId = bundle.getString("roomId")!!,
                callerId = bundle.getString("callerId")!!,
                callerName = bundle.getString("callerName")!!,
                receiverId = bundle.getString("receiverId")!!,
                type = bundle.getString("type")!!,
                avatarUrl = bundle.getString("avatarUrl"),
                mediaType = bundle.getString("mediaType"),
                metadata = metadata,
            )
        }
    }
}

